pkg/shred/shred.go

```diff
@@ -79,6 +84,41 @@ func (d *DataV2Header) LastInSlot() bool {
 type Entry struct {
 	NumHashes uint64
 	Hash      solana.Hash
-	NumTxns   uint64 `bin:"sizeof=Txns"`
+	NumTxns   uint64
 	Txns      []solana.Transaction
 }
+
+func (en *Entry) UnmarshalWithDecoder(decoder *bin.Decoder) (err error) {
+	// read the number of hashes:
+	if en.NumHashes, err = decoder.ReadUint64(bin.LE); err != nil {
+		return fmt.Errorf("failed to read number of hashes: %w", err)
+	}
+	// read the hash:
+	{
+		_, err := decoder.Read(en.Hash[:])
+		if err != nil {
+			return fmt.Errorf("failed to read hash: %w", err)
+		}
+	}
+	// read the number of transactions:
+	{
+		numTxns, err := decoder.ReadUint64(bin.LE)
+		if err != nil {
+			return fmt.Errorf("failed to read number of transactions: %w", err)
+		}
+		en.NumTxns = uint64(numTxns)
+	}
+	if en.NumTxns == 0 {
+		en.Txns = make([]solana.Transaction, 0)
+		return nil
+	}
+	// read the transactions:
+	en.Txns = make([]solana.Transaction, en.NumTxns)
```

Comment:
unsafe slice allocation (length from user-provided input)

Reply:
Added a check against the number of bytes remaining.